contracts/src/dice.cpp

```diff
@@ -2,73 +2,76 @@
+    eosio::check(get_param_value(ses_id, constant::min_bet_param_type) != std::nullopt, "absent min bet param");
+    eosio::check(get_param_value(ses_id, constant::max_bet_param_type) != std::nullopt, "absent max bet param");
+    eosio::check(get_param_value(ses_id, constant::max_payout_param_type) != std::nullopt, "absent max payout param");
 }
 
-void dice::check_bet(uint64_t ses_id) {
+void dice::check_bet(uint64_t ses_id) const {
+    const auto min_bet = asset(*get_param_value(ses_id, constant::min_bet_param_type), core_symbol);
+    const auto max_bet = asset(*get_param_value(ses_id, constant::max_bet_param_type), core_symbol);
+
     const auto& session = get_session(ses_id);
-    auto min_bet = asset(*get_param_value(ses_id, min_bet_param_type), core_symbol);
-    auto max_bet = asset(*get_param_value(ses_id, max_bet_param_type), core_symbol);
     eosio::check(min_bet <= session.deposit, "deposit less than min bet");
     eosio::check(max_bet >= session.deposit, "deposit greater than max bet");
 }
 
-asset dice::calc_max_win(uint64_t ses_id, uint32_t num) {
-    const auto& session = get_session(ses_id);
-    auto max_profit = asset(*get_param_value(ses_id, max_payout_param_type), core_symbol);
-
-    auto all_range = 100;
-    auto win_chance = all_range - num;
-
-    if (win_chance > 98) // if player choose '1' then profit -> 0
-        return zero_asset;
+void dice::check_action_params(const std::vector<uint32_t>& params) const {
+    eosio::check(params.size() == 1, "params amount should be 1");
+    eosio::check(params[0] > 0, "number should be more than 0");
+    eosio::check(params[0] < 100, "number should be less than 100");
+}
 
-    return max_profit / win_chance; // 0..max_payout
+double dice::get_win_coefficient(uint32_t num) {
+    return (constant::all_range / (constant::all_range - num)) * (1. - constant::house_edge);
 }
 
 void dice::on_new_game(uint64_t ses_id) {
     check_params(ses_id);
     check_bet(ses_id);
 
-    require_action(ses_id, roll_action_type);
+    require_action(ses_id, constant::roll_action_type);
 }
 
 void dice::on_action(uint64_t ses_id, uint16_t type, std::vector<uint32_t> params) {
-    eosio::check(type == roll_action_type, "allowed only roll action with type 0");
-    eosio::check(params.size() == 1, "params amount should be 1");
-    eosio::check(params[0] > 0, "number should be more than 0");
-    eosio::check(params[0] < 100, "number should be less than 100");
+    eosio::check(type == constant::roll_action_type, "allowed only roll action with type 0");
+    check_action_params(params);
 
-    rolls.emplace(get_self(), [&](auto& row){
+    const auto& number = params[0];
+    rolls.emplace(get_self(), [&](auto& row) {
         row.ses_id = ses_id;
-        row.number = params[0];
+        row.number = number;
     });
 
-    update_max_win(ses_id, calc_max_win(ses_id, params[0]));
+    update_max_win(ses_id, asset(get_win_coefficient(number), core_symbol));
```

Comment:
mb here should be `get_win_payout` instead of `get_win_coefficient` ?

Reply:
I ask you about it, you tell me use relative coefficient. I can change it to absolute value.